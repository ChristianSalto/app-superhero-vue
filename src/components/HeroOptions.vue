<template>
  <div class="options-container">
    <ul>
      <li
        v-for="hero in heros"
        :key="hero.id"
        @click="$emit('onSelect', hero.id)"
      >
        {{ hero.name }}
      </li>
    </ul>
  </div>
</template>

<script>
export default {
  props: {
    heros: {
      type: Array,
      requiered: true,
    },
  },
};
</script>
<style scoped>
ul {
  list-style-type: none;
  padding: inherit;
}
li {
  background-color: white;
  border-radius: 5px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  cursor: pointer;
  margin-bottom: 10px;
  width: 250px;
  padding: 10px;
}

li:hover {
  background-color: rgba(0, 0, 0, 0.05);
}

.options-container {
  display: flex;
  justify-content: center;
}
</style>
