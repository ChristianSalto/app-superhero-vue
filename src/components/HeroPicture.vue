<template>
  <div class="wrapper-img">
    <img
      :class="showHero ? 'show-hero' : 'hidden-hero'"
      :src="imgSrc"
      alt="hero"
    />
  </div>
</template>

<script>
export default {
  props: {
    image: { type: String, required: true },
    showHero: {
      type: Boolean,
      required: true,
      default: false,
    },
  },
  //   data(){
  //     return this.heroId
  //   },
  computed: {
    imgSrc() {
      return this.image;
    },
  },
};
</script>

<style scoped>
.wrapper-img {
  padding: 2rem;
}

.hidden-hero {
  filter: blur(50px);
}

.show-hero {
  filter: blur(0);
}
</style>

