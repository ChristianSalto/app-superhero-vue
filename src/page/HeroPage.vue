<template>
  <h1 v-if="!hero">Espere por favor...</h1>
  <div v-else>
    <h1>Â¿Quien es este heroe?</h1>

    <HeroPicture
      :image="hero.image"
      :showHero="hero.id === heroId ? !showHero : showHero"
    />
    <HeroOptions :heros="herosArr" @on-select="checkAnswer($event)" />
  </div>

  <template v-if="showAnswer">
    <h2 v-html="msg" class="fade-in"></h2>
    <button @click="newGame">Start game</button>
  </template>
</template>

<script>
import HeroOptions from "@/components/HeroOptions.vue";
import HeroPicture from "@/components/HeroPicture.vue";

import { getHeroOptions } from "@/helpers/getHeroOptions";

export default {
  components: { HeroOptions, HeroPicture },
  data() {
    return {
      herosArr: [],
      hero: null,
      showHero: false,
      heroId: null,
      showAnswer: false,
      msg: "",
    };
  },
  methods: {
    async mixHeroArray() {
      this.herosArr = await getHeroOptions();

      const rndInt = Math.floor(Math.random() * 4);
      this.hero = this.herosArr[rndInt];
    },

    checkAnswer(heroId) {
      this.heroId = heroId;
      this.showAnswer = true;

      if (heroId != this.hero.id) {
        this.msg = `Oops, Incorrecto era <span class="span">${this.hero.name}</span>`;
      } else {
        this.msg = `Correcto es ${this.hero.name}`;
      }
      
      return this.msg;
    },

    newGame() {
      this.showAnswer = false;
      this.showHero = false;
      this.herosArr = [];
      this.mixHeroArray();
    },
  },
  mounted() {
    this.mixHeroArray();
  },
};
</script>

<style scoped>
div h1 {
  color: blue;
}

button {
  padding: 10px;
  border: 0;
  cursor: pointer;
  border-radius: 2px;
  background: whitesmoke;
}

button:hover {
  background: ghostwhite;
}
</style>