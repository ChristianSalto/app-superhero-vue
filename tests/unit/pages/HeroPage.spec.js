import { shallowMount, mount } from '@vue/test-utils'
import HeroPage from '@/page/HeroPage'
import HeroOptions from '@/components/HeroOptions'
import HeroPicture from '@/components/HeroPicture'
import { heros } from '../mocks/heros.mock';

describe('HeroPage Component', () => {

    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(HeroPage)
    })

    test('debe de hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe de llamar mixHeroArray al montar', () => {
        const mixHeroArray = jest.spyOn(HeroPage.methods, 'mixHeroArray')
        const wrapper = shallowMount(HeroPage)

        expect(mixHeroArray).toHaveBeenCalled()
    })

    test('debe de hacer match con el snapshot cuando cargan los pokemons', () => {
        const wrapper = shallowMount(HeroPage, {
            data() {
                return {
                    herosArr: heros,
                    hero: heros[0],
                    showHero: false,
                    heroId: null,
                    showAnswer: false,
                    msg: "",
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe de mostrar los componentes de HeroOptions y HeroPicture', () => {
        const wrapper = shallowMount(HeroPage, {
            data() {
                return {
                    herosArr: heros,
                    hero: heros[0],
                    showHero: false,
                    heroId: null,
                    showAnswer: false,
                    msg: "",
                }
            }
        })

        const heroOptionsComponent = wrapper.findComponent(HeroOptions)
        const heroPictureComponent = wrapper.findComponent(HeroPicture)

        expect( heroOptionsComponent.exists() ).toBe(true)
        expect( heroPictureComponent.exists() ).toBe(true)

        expect( heroPictureComponent.attributes('image') ).toBe('https://www.superherodb.com/pictures2/portraits/10/100/10060.jpg')
        expect( heroOptionsComponent.attributes('heros') ).toBeDefined();
        expect( heroOptionsComponent.attributes('heros') ).toBeTruthy();
    })

    test('pruebas con checkAnswer', async ()=> {
        const wrapper = shallowMount(HeroPage, {
            data() {
                return {
                    herosArr: heros,
                    hero: heros[0],
                    showHero: false,
                    heroId: null,
                    showAnswer: false,
                    msg: "",
                }
            }
        })

        
        await wrapper.vm.checkAnswer(1)

        expect( wrapper.find('h2').exists() ).toBeTruthy()
        expect( wrapper.vm.showAnswer ).toBe(true)
        expect( wrapper.find('h2').text() ).toBe(`Correcto es ${heros[0].name}`)

        await wrapper.vm.checkAnswer(10)

        expect(wrapper.vm.msg).toBe(`Oops, Incorrecto era <span class="span">${heros[0].name}</span>`)
        console.log(wrapper.vm.msg);
    })
})