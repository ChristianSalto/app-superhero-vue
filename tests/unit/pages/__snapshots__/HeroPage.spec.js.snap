// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroPage Component debe de hacer match con el snapshot 1`] = `
<h1>Espere por favor...</h1>
<!--v-if-->
`;

exports[`HeroPage Component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>Â¿Quien es este heroe?</h1>
  <hero-picture-stub image="https://www.superherodb.com/pictures2/portraits/10/100/10060.jpg" showhero="false"></hero-picture-stub>
  <hero-options-stub heros="[object Object],[object Object],[object Object],[object Object]"></hero-options-stub>
</div>
<!--v-if-->
`;
